File: Algorithm-Toolbox-Coursera/week3_greedy_algorithms/3_car_fueling/CarFueling.java
import java.util.*;
import java.io.*;

public class CarFueling {
    static int computeMinRefills(int dist, int tank, int[] stops) {
        int numStops = stops.length;
        int numFills = 0;
        int currFill = 0;
        int lastFillValue = 0;


        if (dist- stops[numStops -1] > tank) {
            return -1;
        }

        while (currFill < numStops) {
            int lastFill = currFill;

            while (currFill < numStops && stops[currFill] - lastFillValue <= tank) {
                currFill = currFill + 1;
            }
            //System.out.print(currFill);
            //System.out.println("fill");

            if (currFill == lastFill) {
                return -1;
            }
            if (dist - lastFillValue <= tank) {
                return numFills;
            }

            if (currFill <= numStops) {
                lastFillValue = stops[currFill - 1];
                numFills += 1;
                }
                
            }
          

        if (dist - lastFillValue > tank) {
            return -1;
        } else {
            return numFills;
        }
        
    
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int dist = scanner.nextInt();
        int tank = scanner.nextInt();
        int n = scanner.nextInt();
        int stops[] = new int[n];
        for (int i = 0; i < n; i++) {
            stops[i] = scanner.nextInt();
        }

        System.out.println(computeMinRefills(dist, tank, stops));
    }
}

File: Algorithm-Toolbox-Coursera/week3_greedy_algorithms/3_car_fueling/car_fueling.py
# python3
import sys


def compute_min_refills(distance, tank, stops):
    stops.append(distance)
    stops.append(0)
    num_stops = len(stops)
    stops.sort()
    #print(stops)
    num_refill = 0
    current_refill = 0
    
    while current_refill < num_stops:
        last_fill = current_refill
        
        while current_refill <= num_stops -1 and (stops[current_refill+1]-stops[last_fill]) <= tank:
            current_refill = current_refill + 1
            #print(last_fill, current_refill)
        if current_refill == last_fill:
            return -1
        if current_refill <= num_stops:
            num_refill = num_refill + 1
  
    
    return num_refill

if __name__ == '__main__':
    d, m, _, *stops = map(int, sys.stdin.read().split())
    print(compute_min_refills(d, m, stops))
    
